src/ceph_exporter.cc

```diff
@@ -0,0 +1,204 @@
+#include "mgr/DaemonMetricCollector.h"
+#include <iostream>
+#include "global/global_init.h"
+#include "common/ceph_argparse.h"
+#include "common/hostname.h"
+#include <map>
+#include <boost/beast/core.hpp>
+#include <boost/beast/http.hpp>
+#include <boost/beast/version.hpp>
+#include <boost/asio.hpp>
+#include <chrono>
+#include <cstdlib>
+#include <ctime>
+#include <iostream>
+#include <memory>
+#include <string>
+
+
+namespace beast = boost::beast;         // from <boost/beast.hpp>
+namespace http = beast::http;           // from <boost/beast/http.hpp>
+namespace net = boost::asio;            // from <boost/asio.hpp>
+using tcp = boost::asio::ip::tcp;       // from <boost/asio/ip/tcp.hpp>
+
+namespace my_program_state
+{
+    std::size_t
+    request_count()
+    {
+        static std::size_t count = 0;
+        return ++count;
+    }
+
+    std::time_t
+    now()
+    {
+        return std::time(0);
+    }
+}
+
+class http_connection : public std::enable_shared_from_this<http_connection>
+{
+public:
+    http_connection(tcp::socket socket)
+        : socket_(std::move(socket))
+    {
+    }
+
+    // Initiate the asynchronous operations associated with the connection.
+    void start()
+    {
+        read_request();
+        check_deadline();
+    }
+
+private:
+    tcp::socket socket_;
+    beast::flat_buffer buffer_{8192};
+    http::request<http::dynamic_body> request_;
+    http::response<http::dynamic_body> response_;
+
+    net::steady_timer deadline_{
+        socket_.get_executor(), std::chrono::seconds(60)};
+
+    // Asynchronously receive a complete request message.
+    void read_request()
+    {
+        auto self = shared_from_this();
+
+        http::async_read(
+            socket_,
+            buffer_,
+            request_,
+            [self](beast::error_code ec,
+                std::size_t bytes_transferred)
+            {
+                boost::ignore_unused(bytes_transferred);
+                if(!ec)
+                    self->process_request();
+            });
+    }
+
+    // Determine what needs to be done with the request message.
+    void process_request()
+    {
+        response_.version(request_.version());
+        response_.keep_alive(false);
```

Comment:
why disable http keepalive here?

as a server, you should generally leave this up to the clients - they'll know whether or not they intend to send more requests on this connection. the clients communicate this in the [Connection](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection) header

you can check whether the client requested keep-alive using `request_.keep_alive()`. if that's true, then you should call back into `read_request()` after `write_response()`'s successful completion to handle the next request on this connection